File: src/ui/updater.rs
//! Background app-update schedule and the install/restart flow
//! (mirrors Go `internal/ui/update_windows.go`).
//!
//! The loop keeps ticking while automatic updates are off (so re-enabling
//! works without a restart), exits permanently on a dev build, and announces
//! any given version at most once.

use std::{sync::atomic::Ordering, thread, time::Duration};

use windows::Win32::{
    Foundation::{LPARAM, WPARAM},
    UI::WindowsAndMessaging::{PostMessageW, WM_CLOSE},
};

use super::{
    imp::{lock, Ui, WM_UPDATE_AVAILABLE},
    task_dialog,
};
use crate::{agent_cli, app_update, version};

const FIRST_CHECK_DELAY: Duration = Duration::from_secs(5);
const CHECK_INTERVAL: Duration = Duration::from_secs(24 * 60 * 60);

/// Starts the periodic update check.
pub(super) fn spawn(ui: &'static Ui) {
    thread::Builder::new()
        .name("app-update".to_owned())
        .spawn(move || {
            let mut offered: Option<String> = None;
            thread::sleep(FIRST_CHECK_DELAY);
            loop {
                if ui.app.automatic_updates_enabled() && !check_once(ui, &mut offered) {
                    return;
                }
                thread::sleep(CHECK_INTERVAL);
            }
        })
        .ok();
}

/// One check; returns `false` when the loop should stop (dev build).
fn check_once(ui: &'static Ui, offered: &mut Option<String>) -> bool {
    match app_update::check(&agent_cli::base_url(), version::VERSION) {
        Err(app_update::Error::DevBuild) => {
            tracing::debug!("development build; update checks disabled");
            false
        }
        Err(err) => {
            tracing::debug!("update check: {err}");
            true
        }
        Ok(None) => true,
        Ok(Some(update)) => {
            if offered.as_deref() != Some(update.version.as_str()) {
                *offered = Some(update.version.clone());
                *lock(&ui.pending_update) = Some(update);
                // SAFETY: posting to our own window from a worker thread.
                unsafe {
                    let _ =
                        PostMessageW(Some(ui.hwnd()), WM_UPDATE_AVAILABLE, WPARAM(0), LPARAM(0));
                }
            }
            true
        }
    }
}

/// Offers the pending update; on acceptance downloads, installs, and prompts
/// for a restart. Re-entry is blocked while an install is in flight.
pub(super) fn offer_install(ui: &'static Ui) {
    if ui.installing.swap(true, Ordering::SeqCst) {
        return;
    }
    thread::Builder::new()
        .name("app-install".to_owned())
        .spawn(move || {
            install_flow(ui);
            ui.installing.store(false, Ordering::SeqCst);
        })
        .ok();
}

fn install_flow(ui: &'static Ui) {
    let Some(update) = lock(&ui.pending_update).clone() else {
        return;
    };

    let accepted = task_dialog::choice(
        &format!("Version {} is available", update.version),
        &format!(
            "You have {}. Tokitoki keeps running while the update downloads and installs.",
            version::VERSION
        ),
        &["Install now", "Not now"],
    );
    if accepted != Some(0) {
        return;
    }

    if let Err(err) = app_update::install(&update) {
        task_dialog::error("Update failed", &err.to_string());
        return;
    }
    lock(&ui.pending_update).take();

    let restart = task_dialog::choice(
        &format!("Version {} is installed", update.version),
        "Restarting takes a few seconds…",
        &["Restart now", "Later"],
    );
    if restart == Some(0) {
        if let Ok(exe) = std::env::current_exe() {
            ui.request_restart(exe);
        }
        // WM_CLOSE → DefWindowProc destroys the window → WM_DESTROY quits the
        // loop; main then relaunches after releasing the instance mutex.
        // SAFETY: posting to our own window from a worker thread.
        unsafe {
            let _ = PostMessageW(Some(ui.hwnd()), WM_CLOSE, WPARAM(0), LPARAM(0));
        }
    }
}

File: src/ui/updater.c
/* Background app-update schedule and the install/restart flow. The loop
 * keeps ticking while automatic updates are off (re-enabling works without
 * a restart), exits permanently on a dev build, and announces any given
 * version at most once. */
#include "ui/ui.h"

#include "agent_cli.h"
#include "app.h"
#include "version.h"

#include <stdio.h>

#define FIRST_CHECK_DELAY_MS (5u * 1000u)
#define CHECK_INTERVAL_MS (24u * 60u * 60u * 1000u)

void updater_set_pending(const UpdateInfo *update) {
    AcquireSRWLockExclusive(&g_ui.lock);
    g_ui.pending_update = *update;
    g_ui.has_pending_update = true;
    ReleaseSRWLockExclusive(&g_ui.lock);
}

/* One check; returns false when the loop should stop (dev build). */
static bool check_once(void) {
    wchar_t base[1024];
    agent_base_url(base, 1024);
    UpdateInfo update;
    UpdateCheck result = app_update_check(base, version_utf8(), &update);
    if (result == UPDATE_DEV_BUILD) {
        return false;
    }
    if (result != UPDATE_AVAILABLE) {
        return true;
    }
    AcquireSRWLockExclusive(&g_ui.lock);
    bool already_offered =
        strcmp(g_ui.offered_version, update.version) == 0;
    if (!already_offered) {
        strcpy_s(g_ui.offered_version, sizeof(g_ui.offered_version), update.version);
        g_ui.pending_update = update;
        g_ui.has_pending_update = true;
    }
    ReleaseSRWLockExclusive(&g_ui.lock);
    if (!already_offered) {
        PostMessageW(g_ui.hwnd, WM_UPDATE_AVAILABLE, 0, 0);
    }
    return true;
}

static DWORD WINAPI updater_main(LPVOID param) {
    (void)param;
    Sleep(FIRST_CHECK_DELAY_MS);
    for (;;) {
        if (app_automatic_updates_enabled() && !check_once()) {
            return 0;
        }
        Sleep(CHECK_INTERVAL_MS);
    }
}

void updater_spawn(void) {
    HANDLE thread = CreateThread(NULL, 0, updater_main, NULL, 0, NULL);
    if (thread) {
        CloseHandle(thread);
    }
}

static DWORD WINAPI install_main(LPVOID param) {
    (void)param;
    AcquireSRWLockShared(&g_ui.lock);
    bool has_update = g_ui.has_pending_update;
    UpdateInfo update = g_ui.pending_update;
    ReleaseSRWLockShared(&g_ui.lock);
    if (!has_update) {
        InterlockedExchange(&g_ui.installing, 0);
        return 0;
    }

    wchar_t instruction[128];
    wchar_t content[256];
    _snwprintf_s(instruction, ARRAYSIZE(instruction), _TRUNCATE,
                 L"Version %hs is available", update.version);
    _snwprintf_s(content, ARRAYSIZE(content), _TRUNCATE,
                 L"You have %s. Tokitoki keeps running while the update "
                 L"downloads and installs.",
                 version_wide());
    static const wchar_t *offer_links[] = {L"Install now", L"Not now"};
    if (task_dialog_choice(instruction, content, offer_links, 2) != 0) {
        InterlockedExchange(&g_ui.installing, 0);
        return 0;
    }

    if (!app_update_install(&update)) {
        task_dialog_error(L"Update failed",
                          L"The update could not be downloaded and verified. "
                          L"Tokitoki keeps running on the current version.");
        InterlockedExchange(&g_ui.installing, 0);
        return 0;
    }

    AcquireSRWLockExclusive(&g_ui.lock);
    g_ui.has_pending_update = false;
    ReleaseSRWLockExclusive(&g_ui.lock);

    _snwprintf_s(instruction, ARRAYSIZE(instruction), _TRUNCATE,
                 L"Version %hs is installed", update.version);
    static const wchar_t *restart_links[] = {L"Restart now", L"Later"};
    if (task_dialog_choice(instruction, L"Restarting takes a few seconds…",
                           restart_links, 2) == 0) {
        wchar_t exe[MAX_PATH];
        if (GetModuleFileNameW(NULL, exe, MAX_PATH)) {
            AcquireSRWLockExclusive(&g_ui.lock);
            wcscpy_s(g_ui.restart_target, MAX_PATH, exe);
            g_ui.restart_requested = true;
            ReleaseSRWLockExclusive(&g_ui.lock);
            /* WM_CLOSE -> DefWindowProc destroys -> WM_DESTROY quits the
             * loop; main relaunches after releasing the instance mutex. */
            PostMessageW(g_ui.hwnd, WM_CLOSE, 0, 0);
        }
    }
    InterlockedExchange(&g_ui.installing, 0);
    return 0;
}

void updater_offer_install(void) {
    if (InterlockedCompareExchange(&g_ui.installing, 1, 0) != 0) {
        return;
    }
    HANDLE thread = CreateThread(NULL, 0, install_main, NULL, 0, NULL);
    if (thread) {
        CloseHandle(thread);
    } else {
        InterlockedExchange(&g_ui.installing, 0);
    }
}
